internal/pkg/service/stream/storage/level/local/writer/writer.go

```diff
@@ -1,42 +1,360 @@
-	UncompressedSize() datasize.ByteSize
+	SliceKey() model.SliceKey
 
-	// WriteRow of tabular data.
-	WriteRow(timestamp time.Time, values []any) error
+	WriteRecord(timestamp time.Time, values []any) error
 	// Close the writer and sync data to the disk.
 	Close(context.Context) error
-
 	// Events provides listening to the writer lifecycle.
 	Events() *Events
 	// DirPath is absolute path to the slice directory. It contains slice file and optionally an auxiliary files.
 	DirPath() string
 	// FilePath is absolute path to the slice file.
 	FilePath() string
 }
+
+type StatisticsProvider interface {
+	// AcceptedWrites returns count of write operations waiting for the sync.
+	AcceptedWrites() uint64
+	// CompletedWrites returns count of successfully written and synced writes.
+	CompletedWrites() uint64
+	// FirstRecordAt returns timestamp of receiving the first record for processing.
+	FirstRecordAt() utctime.UTCTime
+	// LastRecordAt returns timestamp of receiving the last record for processing.
+	LastRecordAt() utctime.UTCTime
+	// CompressedSize written to the file, measured after compression writer.
+	CompressedSize() datasize.ByteSize
+	// UncompressedSize written to the file, measured before compression writer.
+	UncompressedSize() datasize.ByteSize
+}
+
+// writer implements Writer interface, it wraps common logic for all file types.
+// For conversion between record values and bytes, the Writer is used.
+type writer struct {
+	dirPath  string
+	filePath string
+
+	logger log.Logger
+	slice  *model.Slice
+	events *Events
+
+	chain  *writechain.Chain
+	syncer *disksync.Syncer
+
+	formatWriter FormatWriter
+	// closed blocks new writes
+	closed chan struct{}
+	// writeWg waits for in-progress writes before Close
+	writeWg *sync.WaitGroup
+
+	acceptedWrites    *count.Counter
+	completedWrites   *count.CounterWithBackup
+	compressedMeter   *size.MeterWithBackup
+	uncompressedMeter *size.MeterWithBackup
+}
+
+func New(
+	ctx context.Context,
+	logger log.Logger,
+	clk clock.Clock,
+	cfg Config,
+	slice *model.Slice,
+	file writechain.File,
+	dirPath string,
+	filePath string,
+	syncerFactory disksync.SyncerFactory,
+	formatWriterFactory FormatWriterFactory,
+	volumeEvents *Events,
+) (out Writer, err error) {
+	w := &writer{
+		dirPath:  dirPath,
+		filePath: filePath,
+		logger:   logger.WithComponent("slice-writer"),
+		slice:    slice,
+		events:   volumeEvents.Clone(), // clone volume events, to attach additional writer specific events
+		closed:   make(chan struct{}),
+		writeWg:  &sync.WaitGroup{},
+	}
+
+	// Close resources on error
+	defer func() {
+		if err != nil {
+			if w.syncer != nil {
+				_ = w.syncer.Stop(ctx)
+			}
+			if w.chain != nil {
+				_ = w.chain.Close(ctx)
+			}
+			if w.completedWrites != nil {
+				_ = w.completedWrites.Close()
+			}
+		}
+	}()
+
+	// Create counters
+	{
+		// In progress writes counter
+		w.acceptedWrites = count.NewCounter()
```

Comment:
Following code is composition of the writers chain, each part of the chain is imported and also their order.

Small change in parameters/logic can reduce the throughput x times.